core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1292,6 +1137,223 @@ class Block : public lifecycle::StateMachine<Derived>, //
+                if constexpr (gr::traits::port::is_input_v<Port>) {
+                    return port.streamReader().available();
+                } else {
+                    return port.streamWriter().available();
+                }
+            }();
+            if constexpr (std::remove_cvref_t<Port>::kIsSynch) {
+                minSync      = std::max(minSync, port.min_samples);
+                maxSync      = std::min(maxSync, port.max_samples);
+                maxAvailable = std::min(maxAvailable, available);
+            } else {                                 // async port
+                if (available >= port.min_samples) { // ensure that process function is called if at least one async port has data available
+                    hasAsync = true;
+                }
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, std::forward<decltype(ports)>(ports));
+        return std::tuple{minSync, maxSync, maxAvailable, hasAsync};
+    }
+
+    /***
+     * Check the input ports for available samples
+     * @return tuple [nextTag, nextEosTag]
+     */
+    auto
+    getTagPositions() {
+        bool        hasTag                = false;
+        std::size_t nextTag               = std::numeric_limits<std::size_t>::max();
+        std::size_t nextEosTag            = std::numeric_limits<std::size_t>::max();
+        bool        asyncEoS              = false;
+        auto        adjust_for_input_port = [this, &hasTag, &nextTag, &nextEosTag, &asyncEoS]<PortLike Port>(Port &port) {
+            if (port.isConnected()) {
+                if constexpr (std::remove_cvref_t<Port>::kIsSynch) {
+                    // get the tag after the one at position 0 that will be evaluated for this chunk.
+                    // nextTag limits the size of the chunk except if this would violate port constraints
+                    nextTag                               = std::min(nextTag, nSamplesUntilNextTag(port, 1).value_or(std::numeric_limits<std::size_t>::max()));
+                    nextEosTag                            = std::min(nextEosTag, samples_to_eos_tag(port).value_or(std::numeric_limits<std::size_t>::max()));
+                    const gr::ConsumableSpan auto tagData = port.tagReader().get(port.tagReader().available());
+                    hasTag |= !tagData.empty() && tagData[0].index == port.streamReader().position() && !tagData[0].map.empty();
+                } else { // async port
+                    if (samples_to_eos_tag(port).transform([&port](auto n) { return n <= port.min_samples; }).value_or(false)) {
+                        asyncEoS = true;
+                    }
+                }
+            }
+        };
+        for_each_port([&adjust_for_input_port](PortLike auto &port) { adjust_for_input_port(port); }, inputPorts<PortType::STREAM>(&self()));
+        return std::tuple{hasTag, nextTag, nextEosTag, asyncEoS};
```

Comment:
Would it be better if one returns a struct instead of tuple? (see previous comment on getPortLimits())